### DISCUSSION TITLE
`custom` + `variant` is throwing some error.

### DISCUSSION THREAD

**Discussion by premdasvm** [open-circle/valibot#553] [Q&A]
Hi everyone,

i've few schemas in my project as shown below,

```ts

const QuestionSchema = v.object(
	{
		Question: v.string(),
		QuestionLevel: v.enum_(QuestionLevel),
		Metadata: v.object({
			Curriculum: v.object({ Name: v.string(), Id: v.number() }),
			Grade: v.object({ Name: v.string(), Id: v.number() }),
			Qualification: v.array(v.object({ Name: v.string(), Id: v.number() })),
			Subjects: v.array(v.object({ Id: v.number(), Name: v.string() }), [v.minLength(1)]),
			Modules: v.array(v.object({ Id: v.number(), Name: v.string() }), [v.minLength(1)]),
			Chapters: v.array(v.object({ Id: v.number(), Name: v.string() }), [v.minLength(1)]),
			Topics: v.array(v.object({ Id: v.number(), Name: v.string() }), [v.minLength(1)]),
		}),
	},
	[v.custom(input => validateMetadata(input), "Subject should'nt be empty")],
);

const validateMetadata = (input: v.Input<typeof QuestionSchema>): boolean => {
	if (input.QuestionLevel === QuestionLevel.SUBJECT && input.Metadata.Subjects.length <= 0) {
		return false;
	}
	return true;
};

const MCQMergedSchema = v.merge([
	QuestionSchema,
	v.object({
		Type: v.literal(QuestionTypes.MULTI_CHOICE),
		Options: v.array(
			v.object({
				Id: v.number(),
				Option: v.nullish(v.string()),
				OptionImage: v.nullish(v.string()),
				CorrectAnswer: v.boolean(),
			}),
			[v.minLength(1)],
		),
		CorrectAnswerCount: v.number(),
	}),
]);

const EssayMergedSchema = v.merge([
	QuestionSchema,
	v.object({
		Type: v.literal(QuestionTypes.ESSAY),
		MinimumWords: v.number(),
		MaximumWords: v.number(),
		AnswerKeys: v.array(v.string()),
	}),
]);
const SPMergedSchema = v.merge([
	QuestionSchema,
	v.object({
		Type: v.literal(QuestionTypes.SUB_PART),
		Questions: v.lazy(() => v.array(QuestionValibotSchema)),
	}),
]);

export const QuestionValibotSchema = v.variant("Type", [
	MCQMergedSchema,
	EssayMergedSchema,
	SPMergedSchema,
]);
```

so basically, I have a common schema with common fields and other schemas merged with the common schema.
The thing I'm trying to achieve is, that if the `QuestionLevel` = `SUBJECT` then `Metadata.Subjects` shouldn't be empty or null, but if `QuestionLevel = "MODULE"` then `Metadata.Subjects` can be empty or null, but then `Metadata.Modules` shouldn't be.

adding the `[v.custom(input => validateMetadata(input), "Subject shouldn't be empty")]` to my common schema is throwing an error in my `variant` method,

![Screenshot 2024-04-29 at 12 41 11 AM](https://github.com/fabian-hiller/valibot/assets/14838550/b9b9a928-9343-43da-9222-cd99e9b1784c)

how should i approach this issue? any help would be much appreciated

**Comment by fabian-hiller** [comment]
Can you send me your code with the enums so that I can run it in our [playground](https://valibot.dev/playground/)?

**Reply by premdasvm** [reply]
Sure,

```ts
export enum QuestionLevel {
	SUBJECT = "SUBJECT",
	MODULE = "MODULE",
	CHAPTER = "CHAPTER",
	TOPIC = "TOPIC",
}
```

```ts
export enum QuestionTypes {
	MULTI_CHOICE = "MULTI_CHOICE",
	FILL_IN_THE_BLANKS = "FILL_IN_THE_BLANKS",
	ESSAY = "ESSAY",
	SINGLE_LINE = "SINGLE_LINE",
	MULTI_LINE = "MULTI_LINE",
	SUB_PART = "SUB_PART",
}
```

**Reply by fabian-hiller** [reply]
Thank you!

**Comment by fabian-hiller** [comment]
There are two problems (check it out in our [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BTAD0ljiYDsBXAgRR5NUMYBC4AZJhiak4Ab3oBIAMoBVAEIApAKIBhACpwAvHABE67foNmANMoCyAeQAiaiTpPnnbj3eV6ABIAggAKBjoASl5mQWERkf5KBk6hAJJ6MSnpev4Avoys7PDcfHCCwqLiBgCeYMIKju4GaQD6QU4ZnqZmDs1tHV1JAGJpEhKtaQByrQaBOq0aEsFTANIqMaPjkzNzC0sr60k6KirBAJoxJ2fnSSrTAOIerRLT3eb3U08Lr1M6SX0JC0Xm8YoDgb9-vZVJpWqFgpEjD1LHCETZ7AVGABjcQicpCERiLgqLEACyYIGQXgwADoIBQAFZMLEwAAUykUSiUFUJ4gAXJgaSIoMAuABzVkASmh3IJVUk0lkAtppRArVZPPlUhkpGlyiUDiYMGQABNkMblXTGcy2ZyuXoeFARVieKQ+Jb6UyWaz5HApsgQExLcLRRLpXA0ibLbwQBQmFApXA8nquUoHlBTUHBZ6bT6-QGs7SQ+KpbYI1HBTG4wnJUmU1zBGRgNhgFjzUTLcgncgaqzaTnvb7-YHgzARSXw5Ho3xq4nk-XVDxrSzUJ3u73+8vbeXp7H46X8yPBcWw3WywBtWkgUVScUwUmsgCMkoAuguHBATa7hGuMxurV625TpWM77uGw6FkKY6hnO4aXjS14KneD7Pm+MpKHopLIGAMDxqugpdn+fYAbmvrAbSVZgWWEGjuOp7JheV43twYr3k+r4LgYEBgK2+G0oRPbEQOQEVhRoE1tRBa0TBtYMXA8GIberEoRxMoMcoeTQgA9Fp8EuiI+CsqKYA8PAxgAHyYE2Zq4Yaxo2cgRlcCZMDhhYS6AXAqCkhAromoQXDwHGnDgDANRmK+9iSgw9A4lweIkOQNlMHZprmlSphOS5lppM5pkADxhfUEDYPilREiS5KUuZkoClQECkEwyBcCYlmcs2cBZaZNKakS2qyCYximL14j9aQNKWLohhwAAZDNcDGd1qUORNHk2qgNKNchcD5aYAAMtZ2lARqOi12BkKgTAMEoBRKMdMCnXAY5CAweQxXFeIOHo-CGlAYpMCalUUhlgqBn9TCsueygjcSZLA9Cm6AT6+q1PUlrkLhGakBqcpEqjwg0uCAyBJ0eg6PWTg4USfE0gJvb6koiOkQzSjkTSlESSzlPytOpDkN5xEnlKC5KNzRJpJS-28-zD5FtBE4i3o0DHSywTxQA7vGlr1Y1zWlgz6mpgpzHIexL4yvWStOjaauoJrUBKzwgW7rOKbqW+724vAOioKgPa-f9gNw5S1IIfG-2Q9DuPiEDlIIyRg4o3UkEY-GZA4+V1TJxt1wXPWDiiqAfAAOrQCaNMc-rXIOMgLBFyApdQOXLtUfqtv26sTA1DTdOC-LYZu3qHuxV7cAqKEAcA7HINXuHENQ7Kmew1VyDx8JyNcvj6PAJj6cw-jG0ovCiL1jpMM06QyAAF69omFkEeuGe8lwABqTZUDA0+SoP9jD8wbDQHgB9eAMM37kA-tPUOJARTNTZGYfGdh5KOG+pPIOK9oQ+z9jUVB09oTjxwcHVe9APZAA)). The first is the pipe of `QuestionSchema` because it references itself. This seems to be a TypeScript problem. Valibot can't do anything about it. Also, the pipeline is ignored when you merge `QuestionSchema` with other objects because properties may be over-written, making the pipeline typing invalid. We will improve the DX in our next release.

The same problem occurs when using `lazy`. Due to a TypeScript limitation, the input and output types of `lazy` schemas cannot be inferred automatically. Therefore, you must explicitly specify these types using [`BaseSchema`](https://valibot.dev/api/BaseSchema/). Read more [here](https://valibot.dev/guides/other/#lazy-schema).
